py/kubeflow/testing/new_test_util.py

```diff
@@ -0,0 +1,155 @@
+"""
+This module contains misc utils used for Kubeflow testing
+infrastructure. They include creating test cases/suites,
+managing logs of test cases and initializing tests
+
+This is intended to replace test_util.py
+"""
+import argparse
+import logging
+import os
+import subprocess
+import time
+import xml.etree.ElementTree as ET
+
+import junit_xml
+
+
+class TestCase(junit_xml.TestCase):
+  """A single test case."""
+
+  def __init__(self, test_func, **kwargs):
+    self.class_name = "Kubeflow"
+    self.start_time = time.time()
+    self.test_func = test_func
+    super(TestCase, self).__init__(**kwargs)
+
+  def add_failure_info(self, message=None, output=None, failure_type=None):
+    if message:
+      if not self.failure_message:
+        self.failure_message = ''
+      self.failure_message += '\n' + message
+    if output:
+      if not self.failure_output:
+        self.failure_output = ''
+      self.failure_output += '\n' + output
+    if failure_type:
+      self.failure_type = failure_type
+
+class TestSuite(junit_xml.TestSuite):
+  """A suite of test cases."""
+
+  def __init__(self, name, artifacts_dir, logs_dir, **kwargs):
+    self.artifacts_dir = artifacts_dir
+    self.logs_dir = logs_dir
+    super(TestSuite, self).__init__(name, **kwargs)
+
+  def generate_xml(self):
+    output_file = os.path.join(self.artifacts_dir,
+                               "junit_" + self.name + ".xml")
+    ET.ElementTree(ET.fromstring(junit_xml.TestSuite.to_xml_string(
```

Comment:
what's this elementTree doing?

Reply:
My bad. This long chained function call is not easy to understand. Added comment to explain what's happening and broke this down into 3 separate calls.